### PR TITLE
add support for client cert auth for mysql protocol vtgate connections

### PR DESCRIPTION
**Pull request by tpetr**
This PR adds a new vtgate auth plugin (`clientcert`) which populates the `VTGateCallerID` from the CommonName and DNSNames values in the connection's client certificate (similar to gRPC).

### CODE REVIEW

File: go/mysql/clientcert/auth_server_clientcert.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+func (ascc *AuthServerClientCert) ValidateHash(salt []byte, user string, authResponse []byte, remoteAddr net.Addr) (mysql.Getter, error) {
+	panic("unimplemented")
+}
+
+// Negotiate is part of the AuthServer interface.
+func (ascc *AuthServerClientCert) Negotiate(c *mysql.Conn, user string, remoteAddr net.Addr) (mysql.Getter, error) {
+	// This code depends on the fact that golang's tls server enforces client cert verification.
+	// Note that the -mysql_server_ssl_ca flag must be set in order for the vtgate to accept client certs.
+	// If not set, the vtgate will effectively deny all incoming mysql connections, since they will all lack certificates.
+	// For more info, check out go/vt/vtttls/vttls.go
```
Comment (tpetr) [inline go/mysql/clientcert/auth_server_clientcert.go]: I should probably just make this idiot-proof by `panic()`-ing in `Init()` if `-mysql_server_ssl_ca` is not set...

File: go/mysql/clientcert/auth_server_clientcert.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	certs := c.GetTLSClientCerts()
+	if certs == nil || len(certs) == 0 {
```
Comment (dweitzman) [inline go/mysql/clientcert/auth_server_clientcert.go]: The nil check should be unnecessary since len(nil) == 0

**Comment by sougou** [comment]
@tpetr I ( or @dweitzman ) can merge this once you address the review comment.

**Comment by tpetr** [comment]
Great, thanks. Apologies for the delay -- will address the feedback sometime this week.

**Comment by tpetr** [comment]
This should be good for another look -- I don't think the most recent test failure is due to this PR (is there a way to re-run a travis build?)

**Comment by deepthi** [comment]
> This should be good for another look -- I don't think the most recent test failure is due to this PR (is there a way to re-run a travis build?)

I re-ran the job that failed.
If you are logged into Travis CI you should be able to restart a failed build.

**Comment by tpetr** [comment]
@sougou @dweitzman can you take another peek at this when you have a minute?

**Comment by sougou** [comment]
@dkhenry

**Comment by tpetr** [comment]
@dkhenry could I get some 👀? Thanks!

File: go/mysql/auth_server_clientcert.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	commonName := certs[0].Subject.CommonName
+
+	if user != commonName {
+		return nil, fmt.Errorf("MySQL connection username '%v' does not match client cert common name '%v'", user, commonName)
+	}
+
+	return &StaticUserData{
+		username: commonName,
+		groups:   certs[0].DNSNames,
```
Comment (dkhenry) [inline go/mysql/auth_server_clientcert.go]: This is fine for this implementation, but it should be moved to use Subject Alternative Names

**Pull request merged**